### PR TITLE
Add built in metric ndcg_at_k to retriever evaluation

### PR DESCRIPTION
**Pull request by liangz1**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/liangz1/mlflow/pull/10284?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10284/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10284
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Add NDCG@k for retriever evaluation.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [x] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Added NDCG@k for retriever evaluation.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/metrics/metric_definitions.py [hunk 1]
```diff
@@ -373,6 +373,52 @@ def _fn(predictions, targets):
     return _fn
 
 
+def _prepare_for_ndcg(predictions, targets, k):
+    """Prepare data from predictions and targets to y_score, y_true for ndcg calculation.
+
+    Args:
+        predictions: A list of tuples of strings of doc IDs.
+        targets: A list of tuples of strings of doc IDs.
```
Comment (dbczumar) [inline mlflow/metrics/metric_definitions.py:381]: I think @bbqiu converted these to lists

File: mlflow/metrics/metric_definitions.py [hunk 2]
```diff
@@ -373,6 +373,94 @@ def _fn(predictions, targets):
...
+    # support predictions contain duplicate doc ID
+    predictions, targets = _expand_duplicate_retrieved_docs(predictions, targets)
+
+    all_docs = targets.union(predictions)
+    doc_id_to_index = {doc_id: i for i, doc_id in enumerate(all_docs)}
+    n_labels = max(len(doc_id_to_index), 2)  # sklearn.metrics.ndcg_score requires at least 2 labels
+    y_true = np.zeros((1, n_labels), dtype=np.float32)
+    y_score = np.zeros((1, n_labels), dtype=np.float32)
+    for i, doc_id in enumerate(predictions):
+        # "1 - i * eps" means we assign higher score to docs that are ranked higher,
```
Comment (sunishsheth2009) [inline mlflow/metrics/metric_definitions.py:419]: How do we know if the docs are ranked higher? We don't store some count anywhere as far as I can tell. Let me know if I am missing something

**Review by sunishsheth2009 — APPROVED** [review]

File: mlflow/metrics/metric_definitions.py (hunk 2 above)
Comment (liangz1) [reply mlflow/metrics/metric_definitions.py:419]: 1. The retrieved rank of the doc is represented by the order in the list in  `predictions` column.
2. The ground truth does not distinguish the rank of the docs. Our ndcg score views all documents in the `targets` column as an unordered set since we use binary relevance (for simplicity).

File: tests/metrics/test_metric_definitions.py [hunk 3]
```diff
@@ -266,3 +270,69 @@ def test_recall_at_k():
         "p90": 0.8,
         "variance": 0.1,
     }
+
+
+def test_ndcg_at_k():
+    # normal cases
```
Comment (bbqiu) [inline tests/metrics/test_metric_definitions.py:276]: nit: I think that some of these cases are covered within `test_default_evaluator.py` already, so it could be worthwhile to trim down on these tests to make it easier to maintain in the future?

File: docs/source/python_api/mlflow.metrics.rst [hunk 4]
```diff
@@ -160,6 +160,7 @@ Parameters:
 
 .. autofunction:: mlflow.metrics.recall_at_k
 
+.. autofunction:: mlflow.metrics.ndcg_at_k
```
Comment (bbqiu) [inline docs/source/python_api/mlflow.metrics.rst:163]: I think you'll have to add to the `Retriever Metrics` blurb above for ndcg as well!
ex. [linked here](https://github.com/mlflow/mlflow/pull/10144/files#diff-ef4ee65c3d24eaada583650c0f1004db2c67b688c0249cf754df6e65dfd0aa7eR145-R155)

File: tests/evaluate/test_default_evaluator.py [hunk 5]
```diff
@@ -3409,6 +3429,9 @@ def fn3(X):
         "recall_at_3/mean": 0.5,
         "recall_at_3/p90": 0.5,
         "recall_at_3/variance": 0.0,
+        "ndcg_at_3/mean": 0.6131471927654585,
+        "ndcg_at_3/p90": 0.6131471927654585,
+        "ndcg_at_3/variance": 0.0,
```
Comment (bbqiu) [inline tests/evaluate/test_default_evaluator.py:3434]: I think we may want to update `test_evaluate_retriever_builtin_metrics_no_model_type` as well

File: mlflow/metrics/metric_definitions.py [hunk 6]
```diff
@@ -373,6 +373,94 @@ def _fn(predictions, targets):
...
+        )
+        return noop
+
+    def _fn(predictions, targets):
+        from sklearn.metrics import ndcg_score
+
+        if not _validate_list_str_data(
+            predictions, "ndcg_at_k", predictions_col_specifier
+        ) or not _validate_list_str_data(targets, "ndcg_at_k", targets_col_specifier):
+            return noop
```
Comment (bbqiu) [inline mlflow/metrics/metric_definitions.py:441]: do we want to return noop or None?

File: mlflow/metrics/metric_definitions.py [hunk 7]
```diff
@@ -373,6 +373,94 @@ def _fn(predictions, targets):
...
+
+        scores = []
+        for i in range(len(predictions)):
```
Comment (bbqiu) [inline mlflow/metrics/metric_definitions.py:444]: nit: can we use something like
`for target, prediction in zip(targets, predictions):`

File: mlflow/metrics/metric_definitions.py [hunk 8]
```diff
@@ -373,6 +373,94 @@ def _fn(predictions, targets):
...
+            if len(predictions[i]) == 0 and len(targets[i]) == 0:
+                scores.append(1)  # no error is made
+                continue
+
+            if len(predictions[i]) == 0 or len(targets[i]) == 0:
+                scores.append(0)
+                continue
+
+            # only include the top k retrieved chunks
+            y_score, y_true = _prepare_row_for_ndcg(predictions[i][:k], targets[i])
```
Comment (bbqiu) [inline mlflow/metrics/metric_definitions.py:455]: nit: can we rename this to `relevance_scores` and `ideal_relevance_scores`? just to make the names a little more descriptive

File: mlflow/metrics/metric_definitions.py (hunk 7 above)
Comment (bbqiu) [reply mlflow/metrics/metric_definitions.py:444]: or something like ground truth and retrieved

File: mlflow/metrics/metric_definitions.py [hunk 9]
```diff
@@ -373,6 +373,94 @@ def _fn(predictions, targets):
...
+        predictions: A list of strings of at most k doc IDs retrieved.
+        targets: A list of strings of ground-truth doc IDs.
+
+    Returns:
+        y_true : ndarray of shape (1, n_docs) Representing the ground-truth relevant docs.
+            n_docs is the number of unique docs in union of predictions and targets.
+        y_score : ndarray of shape (1, n_docs) Representing the retrieved docs.
+            n_docs is the number of unique docs in union of predictions and targets.
+    """
+    eps = 1e-6
```
Comment (bbqiu) [inline mlflow/metrics/metric_definitions.py:407]: could we add a comment here to explain that sklearn does an internal sort of yscore, so to preserve the order of our retrieved docs, you need to modify the relevance value slightly

File: mlflow/metrics/metric_definitions.py (hunk 8 above)
Comment (liangz1) [reply mlflow/metrics/metric_definitions.py:455]: I think changing the `for ground_truth, retrieved in zip(targets, predictions):` will help.
Here, I prefer to keep them `y_score, y_true` so that they match the parameter names in sklearn. The semantics should not be confusing since it's align with sklearn's definition. Wdyt?

File: tests/metrics/test_metric_definitions.py (hunk 3 above)
Comment (liangz1) [reply tests/metrics/test_metric_definitions.py:276]: Discussed offline: for future custom metrics, the correctness of the calculation is preferred to be written in `test_metric_definitions`. `test_default_evaluator` is supposed to make sure the built-in metrics are correctly associated with the corresponding model type.

File: docs/source/python_api/mlflow.metrics.rst (hunk 4 above)
Comment (liangz1) [reply docs/source/python_api/mlflow.metrics.rst:162]: Updated above

File: mlflow/metrics/metric_definitions.py [hunk 10]
```diff
@@ -373,6 +373,94 @@ def _fn(predictions, targets):
...
+        from sklearn.metrics import ndcg_score
+
+        if not _validate_list_str_data(
+            predictions, "ndcg_at_k", predictions_col_specifier
+        ) or not _validate_list_str_data(targets, "ndcg_at_k", targets_col_specifier):
+            return noop
+
+        scores = []
+        for i in range(len(predictions)):
+            # edge cases
```
[the remark this answers is not in the document]
Comment (liangz1) [reply mlflow/metrics/metric_definitions.py]: Added more comments!

File: mlflow/metrics/metric_definitions.py (hunk 6 above)
Comment (liangz1) [reply mlflow/metrics/metric_definitions.py]: Fixed with `return`.

File: docs/source/python_api/mlflow.metrics.rst [hunk 11]
```diff
@@ -146,20 +146,20 @@ Parameters:
...
-                mlflow.metrics.recall_at_k(5)
+                mlflow.metrics.recall_at_k(5),
+                mlflow.metrics.ndcg_at_k(5)
             ]   
         )
     
     NOTE: In the 2nd method, it is recommended to omit the ``model_type`` as well, or else 
     ``precision@3`` and ``recall@3`` will be  calculated in  addition to ``precision@5``, 
-    ``precision@6``, ``recall@4``, and ``recall@5``.
+    ``recall@5``, and ``ndcg_at_k@5``.
```
Comment (bbqiu) [inline docs/source/python_api/mlflow.metrics.rst:156]: ndcg@3 will be calculated by default as well

maybe to match the previous format, you can write ndcg@5?

I also originally wanted this to show how you can use diff values of k, so maybe also calculate precision@6 along with these 3 @5 metrics

File: mlflow/metrics/metric_definitions.py (hunk 8 above)
Comment (bbqiu) [reply mlflow/metrics/metric_definitions.py]: sounds good!

File: mlflow/metrics/metric_definitions.py [hunk 12]
```diff
@@ -373,6 +373,100 @@ def _fn(predictions, targets):
...
+    Returns:
+        y_true : ndarray of shape (1, n_docs) Representing the ground-truth relevant docs.
+            n_docs is the number of unique docs in union of predictions and targets.
+        y_score : ndarray of shape (1, n_docs) Representing the retrieved docs.
+            n_docs is the number of unique docs in union of predictions and targets.
+    """
+    # sklearn does an internal sort of y_score, so to preserve the order of our retrieved
+    # docs, we need to modify the relevance value slightly
+    eps = 1e-6
+    targets = set(targets)
```
Comment (bbqiu) [inline mlflow/metrics/metric_definitions.py:410]: nit: only need either this line or [this line](https://github.com/mlflow/mlflow/pull/10284/files#diff-4d216618e4891c4fb5751f7698964f0ea2fcffb184fd915877e90be401f538c6R379) because both will create a new set without mutating the underlying array

**Review by bbqiu — APPROVED** [review]

File: docs/source/python_api/mlflow.metrics.rst (hunk 11 above)
Comment (liangz1) [reply docs/source/python_api/mlflow.metrics.rst:156]: Got it, sgtm!

File: mlflow/metrics/metric_definitions.py (hunk 12 above)
Comment (liangz1) [reply mlflow/metrics/metric_definitions.py:410]: Oh, nice catch, I remember we discussed this.
